pkg/httpclient/client.go

```diff
@@ -0,0 +1,202 @@
+// Then, to make a request, call one of the service methods, e.g.:
+//		resp := client.GetJSON("http://site/path")
+//
+// Once you have an user and a corresponding public API key, you can issue authenticated requests,
+// by constructing a new client with the appropriate credentials:
+//
+//		client := httpclient.NewClientWithAuthentication(username, publicAPIToken)
+//
+package httpclient
+
+import (
+	"fmt"
+	"github.com/mongodb-labs/pcgc/pkg/useful"
+	"gopkg.in/errgo.v1"
+	"io"
+	"net"
+	"net/http"
+	"runtime"
+
+	digest "github.com/Sectorbob/mlab-ns2/gae/ns/digest"
+)
+
+// ContentTypeJSON defines the JSON content type
+const ContentTypeJSON = "application/json; charset=UTF-8"
+
+// PreferJSON signal that we are accepting JSON responses, but do not reject non-JSON data
+const PreferJSON = "application/json;q=0.9, */*;q=0.8"
+
+var (
+	// userAgent holds a user agent string which will be passed along with every request
+	userAgent string
+	// the version string
+	version string
+)
+
+func init() {
+	ver := version
+	if ver == "" {
+		// if the version is not passed at build time, flag it as 'unknown'
+		ver = "unknown"
+	}
+
+	userAgent = fmt.Sprintf("pcgc/httpclient-%s (%s; %s)", ver, runtime.GOOS, runtime.GOARCH)
+}
+
+type basicHTTPClient struct {
+	client *http.Client
+	auth   *digest.Transport
+}
+
+// HTTPResponse wrapper for HTTP response objects
+type HTTPResponse struct {
+	Response *http.Response
+	Err      error
+}
+
+// BasicHTTPOperation defines a contract for this client's API
+type BasicHTTPOperation interface {
+	GetJSON(url string) HTTPResponse
+	PostJSON(url string, body io.Reader) HTTPResponse
+	PatchJSON(url string, body io.Reader) HTTPResponse
+	PutJSON(url string, body io.Reader) HTTPResponse
+	Delete(url string) HTTPResponse
+}
+
+// Error implementation for error responses
+func (resp HTTPResponse) Error() string {
+	return resp.Err.Error()
+}
+
+// IsError returns true if the associated error is not nil
+func (resp HTTPResponse) IsError() bool {
+	return resp.Err != nil
+}
+
+// NewClient builds a new HTTP client with default timeouts
+func NewClient() BasicHTTPOperation {
+	return NewClientWithTimeouts(InitTimeouts())
+}
+
+// NewClientWithTimeouts builds a new HTTP client with specified timeouts
+func NewClientWithTimeouts(timeouts *RequestTimeouts) BasicHTTPOperation {
+	return basicHTTPClient{client: &http.Client{
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: timeouts.DialTimeout,
+			}).DialContext,
+			ExpectContinueTimeout: timeouts.ExpectContinueTimeout,
+			IdleConnTimeout:       timeouts.IdleConnectionTimeout,
+			ResponseHeaderTimeout: timeouts.ResponseHeaderTimeout,
+			TLSHandshakeTimeout:   timeouts.TLSHandshakeTimeout,
+		},
+		Timeout: timeouts.GlobalTimeout,
+	}}
+}
+
+// NewClientWithAuthentication builds a new client which can use Digest authentication to make authenticated calls
+func NewClientWithAuthentication(user string, publicAPIToken string) BasicHTTPOperation {
```

Comment:
please note that it's not a `publicAPIToken` but a password, and that if we are sticking to the new API keys then user-> public key, password-> privateKey

Reply:
The following can be used for Authentication:
- (user, password): not sure if it works in OM 4.0
- (user, publicAPIToken): deprecated but still works
- or (publicKey, privateKey): OM 4.2.0+